crates/expressions/src/lib.rs

```diff
@@ -52,6 +52,21 @@ impl ProviderResolver {
         self.resolve_template(template).await
     }
 
+    pub async fn resolve_all(&self, component_id: &str) -> Result<Vec<(String, String)>> {
+        use futures::FutureExt;
+
+        let Some(keys2templates) = self.internal.component_configs.get(component_id) else {
+            return Ok(vec![]);
+        };
+
+        let resolve_futs = keys2templates.iter().map(|(key, template)| {
+            self.resolve_template(template)
+                .map(|r| r.map(|value| (key.to_string(), value)))
+        });
+
+        futures::future::try_join_all(resolve_futs).await
```

Comment:
I am slightly worried about performance - at some point we might want to see if adding some sort of caching mechanism to providers is worth it.

Reply:
We should probably push this down into the provider as e.g. (untested):

```rust
trait Provider {
    async fn get(&self, key: &Key) -> anyhow::Result<Option<String>>;

    async fn get_many(&self, keys: impl IntoIterator<Item = &Key>) -> anyhow::Result<impl IntoIterator<Option<String>>> {
        try_join_all(keys.into_iter().map(|key| self.get(key)).await
    }
}
```

Reply:
I don't think any of our current providers can optimise this beyond running the futures concurrently - all appear to support retrieving only one secret at once - so I propose we defer this.

Reply:
@itowlson instead of adding this function, could we use lann's suggestion from [`get_all`](https://github.com/fermyon/spin/pull/2869/files#r1778592084)